pkg/services/meta/meta.go

```diff
@@ -0,0 +1,125 @@
+	// associated with MPT.
+	rootKey = 0x00
+)
+
+// ContainerLister is a source of actual containers current node belongs to.
+type ContainerLister interface {
+	// List returns node's containers and
+	// any error that does not allow listing.
+	List() (map[cid.ID]struct{}, error)
+}
+
+// Meta handles object meta information received from FS chain and object
+// storages. Chain information is stored in Merkle-Patricia Tries. Full objects
+// index is built and stored as a simple KV storage.
+type Meta struct {
+	l        *zap.Logger
+	stopCh   chan struct{}
+	rootPath string
+
+	netmapH util.Uint160
+	cnrH    util.Uint160
+	cLister ContainerLister
+
+	ws      *rpcclient.WSClient
+	bCh     chan *block.Header
+	objEv   chan *state.ContainedNotificationEvent
+	cnrEv   chan *state.ContainedNotificationEvent
+	epochEv chan *state.ContainedNotificationEvent
+
+	m        sync.RWMutex
+	storages map[cid.ID]*containerStorage
+}
+
+// New makes [Meta].
+func New(l *zap.Logger, cLister ContainerLister, ws *rpcclient.WSClient, containerH, nmH util.Uint160, rootPath string) (*Meta, error) {
+	storagesFS, err := os.ReadDir(rootPath)
+	if err != nil {
+		return nil, fmt.Errorf("read existing container storages: %w", err)
+	}
+	storagesRead := make(map[cid.ID]*containerStorage)
+	for _, s := range storagesFS {
+		sName := s.Name()
+		cID, err := cid.DecodeString(sName)
+		if err != nil {
+			l.Warn("skip unknown container storage entity", zap.String("name", sName), zap.Error(err))
+			continue
+		}
+
+		storagesRead[cID], err = storageForContainer(rootPath, cID)
+		if err != nil {
+			l.Warn("skip container storage that cannot be read", zap.String("name", sName), zap.Error(err))
+			continue
+		}
+	}
+
+	cnrsNetwork, err := cLister.List()
+	if err != nil {
+		return nil, fmt.Errorf("listing node's containers: %w", err)
+	}
+	for cID := range storagesRead {
+		if _, ok := cnrsNetwork[cID]; !ok {
+			err = storagesRead[cID].drop()
+			if err != nil {
+				l.Warn("could not drop container storage", zap.Stringer("cID", cID), zap.Error(err))
+			}
+
+			delete(storagesRead, cID)
+		}
+	}
+
+	storages := storagesRead
+	for cID := range cnrsNetwork {
+		if _, ok := storages[cID]; !ok {
+			storages[cID], err = storageForContainer(rootPath, cID)
+			if err != nil {
+				return nil, fmt.Errorf("open container storage %s: %w", cID, err)
+			}
+		}
+	}
+
+	return &Meta{
+		l:        l,
+		stopCh:   make(chan struct{}),
+		rootPath: rootPath,
+		netmapH:  nmH,
+		cnrH:     containerH,
+		cLister:  cLister,
+		ws:       ws,
+		bCh:      make(chan *block.Header),
+		objEv:    make(chan *state.ContainedNotificationEvent),
+		cnrEv:    make(chan *state.ContainedNotificationEvent),
+		epochEv:  make(chan *state.ContainedNotificationEvent),
+		storages: storages}, nil
+}
+
+// Run starts notification handling. Must be called only on instances created
+// with [New]. Blocked until context is done.
+func (m *Meta) Run(ctx context.Context) error {
```

Comment:
Why using context here?

Reply:
https://github.com/nspcc-dev/neofs-node/blob/d7407d117b3fba12088c6b803f1e2a9894294fbe/cmd/neofs-node/worker.go#L15-L23 the only our practice reason. i do not like it